lib/msal-core/src/cache/AuthCache.ts

```diff
@@ -0,0 +1,189 @@
+
+        keysToMigrate.forEach((cacheKey, index) => this.duplicateCacheEntry(cacheKey, values[index], storeAuthStateInCookie));
+    }
+
+    private duplicateCacheEntry(newKey: string, value: string, storeAuthStateInCookie?: boolean) {
+        if (value) {
+            this.setItem(newKey, value, storeAuthStateInCookie);
+        }
+    }
+
+    // Prepend msal.<client-id> to each key
+    private generateCacheKey(key: string, addInstanceId: boolean): string {
+        try {
+            JSON.parse(key);
+            return key;
+        } catch (e) {
+            if (key.startsWith(`${CacheKeys.PREFIX}.${this.clientId}`)) {
+                return key;
+            } else {
+                return addInstanceId ? `${CacheKeys.PREFIX}.${this.clientId}.${key}` : `${CacheKeys.PREFIX}.${key}`;
+            }
+        }
+    }
+
+    // add value to storage
+    setItem(key: string, value: string, enableCookieStorage?: boolean): void {
+        super.setItem(this.generateCacheKey(key, true), value, enableCookieStorage);
+
+        if (this.rollbackEnabled) {
+            super.setItem(this.generateCacheKey(key, false), value, enableCookieStorage);
+        }
+    }
+
+    // get one item by key from storage
+    getItem(key: string, enableCookieStorage?: boolean): string {
+        return super.getItem(this.generateCacheKey(key, true), enableCookieStorage);
+    }
+
+    // remove value from storage
+    removeItem(key: string): void {
+        super.removeItem(this.generateCacheKey(key, true));
+        if (this.rollbackEnabled) {
+            super.removeItem(this.generateCacheKey(key, false));
+        }
+    }
+
+    resetCacheItems(): void {
+        const storage = window[this.cacheLocation];
+        let key: string;
+        this.removeAcquireTokenEntries();
+        for (key in storage) {
+            if (storage.hasOwnProperty(key)) {
+                // Check if key contains msal prefix
+                if (key.indexOf(CacheKeys.PREFIX) !== -1) {
+                    // For now, we are clearing all cache items created by MSAL.js
+                    super.removeItem(key);
+                    // TODO: Clear cache based on client id (clarify use cases where this is needed)
+                }
+            }
+        }
+    }
+
+    setItemCookie(cName: string, cValue: string, expires?: number): void {
+        super.setItemCookie(this.generateCacheKey(cName, true), cValue, expires);
+        if (this.rollbackEnabled) {
+            super.setItemCookie(this.generateCacheKey(cName, false), cValue, expires);
+        }
+    }
+
+    getItemCookie(cName: string): string {
+        return super.getItemCookie(this.generateCacheKey(cName, true));
+    }
+
+    getAllAccessTokens(clientId: string, homeAccountIdentifier: string): Array<AccessTokenCacheItem> {
+        const results: Array<AccessTokenCacheItem> = [];
+        let accessTokenCacheItem: AccessTokenCacheItem;
+        const storage = window[this.cacheLocation];
```

Comment:
I think it would be better if `this.cacheLocation` was an actual refer to `localStorage` or `sessionStorage`, so that you don't need to have this line everywhere (and that's also where you validate if the cache location is valid).

Reply:
Can we do that as a separate PR? We can target solving the type widening errors seen in specific environments too at the same time.

Reply:
Resolving, will address this as a separate PR.